tests/rptest/tests/workload_upgrade_runner_test.py

```diff
@@ -0,0 +1,235 @@
+        rest_versions = self.upgrade_steps[1:]
+        self.logger.info(f"installing base version {current_version}")
+        self.installer.install(nodes=self.redpanda.nodes,
+                               version=current_version)
+        self.redpanda.start(clean_nodes=True)
+
+        installed_logical_version = self.cluster_version()
+        # upgrade loop: for each version
+        for next_version in rest_versions:
+            # setup workload that could start at current_version
+            for w in self.adapted_workloads:
+                if w.state != WorkloadAdapter.NOT_STARTED:
+                    continue
+
+                if current_version >= w.get_earliest_applicable_release():
+                    self.logger.info(f"setup {w.get_workload_name()}")
+                    w.begin()  # this will set in a STARTED state
+
+            # run checks on all the started workload.
+            # each check could take multiple runs, so loop on a list of it until exhaustion
+            self._check_workload_list(to_check_list=[
+                w for w in self.adapted_workloads
+                if w.state == WorkloadAdapter.STARTED
+            ],
+                                      version_param=current_version)
+
+            # stop workload that can't operate with next_version
+            for w in self.adapted_workloads:
+                if w.state != WorkloadAdapter.STARTED:
+                    continue
+
+                if current_version == w.get_latest_applicable_release():
+                    self.logger.info(f"teardown of {w.get_workload_name()}")
+                    w.end()
+
+            # proceed to updgrade nodes to next_version
+            self.logger.info(
+                f"upgrade step for {[n.account.hostname for n in self.redpanda.nodes]} from {current_version=} to {next_version=}"
+            )
+            versions = {n: current_version for n in self.redpanda.nodes}
+            for n in self.redpanda.nodes:
+                # upgrade each node one by one, checking on workload each time
+                self.installer.install(nodes=[n], version=next_version)
+                self.redpanda.rolling_restart_nodes(nodes=[n],
+                                                    start_timeout=90,
+                                                    stop_timeout=90)
+                versions[n] = next_version
+                # check only workload that are active and that can operate with next_version
+                to_check_workloads = [
+                    w for w in self.adapted_workloads
+                    if w.state == WorkloadAdapter.STARTED
+                    and next_version <= w.get_latest_applicable_release()
+                ]
+                self._check_workload_list(to_check_list=to_check_workloads,
+                                          version_param=versions,
+                                          partial_update=True)
+
+            wait_until(
```

Comment:
This is the same logic as `RedpandaTest.upgrade_through_versions`, right?

Can the outer for loop just be replaced with a call to that?  One way or another, we should avoid having two implementations of that logic.